buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -237,22 +232,24 @@ private boolean allocate(int size, int maxCapacity, Thread currentThread, Adapti
             int index = (int) (threadId & mask);
             for (int i = 0, m = Integer.numberOfTrailingZeros(~mask); i < m; i++) {
                 Magazine mag = mags[index + i & mask];
-                if (mag.tryAllocate(size, sizeBucket, maxCapacity, buf)) {
+                AdaptiveByteBuf allocated = mag.tryAllocate(size, sizeBucket, maxCapacity);
+                if (allocated != null) {
                     // Was able to allocate.
-                    return true;
+                    return allocated;
                 }
             }
             expansions++;
         } while (expansions <= EXPANSION_ATTEMPTS && tryExpandMagazines(mags.length));
-        return false;
+        return null;
     }
 
     /**
      * Allocate into the given buffer. Used by {@link AdaptiveByteBuf#capacity(int)}.
      */
     void allocate(int size, int maxCapacity, AdaptiveByteBuf into) {
         Magazine magazine = into.chunk.magazine;
-        if (!allocate(size, maxCapacity, Thread.currentThread(), into)) {
+        int sizeBucket = AllocationStatistics.sizeBucket(size);
+        if (!into.chunk.magazine.tryAllocate(size, sizeBucket, maxCapacity, into)) {
```

Comment:
Please note that with the current implementation we would always stay in the same chunk when trying to increase the capacity.

Reply:
That's not going to work so well, since we pointer-bump allocate from chunks.